authenticatorbridge/src/main/java/com/bitwarden/authenticatorbridge/manager/AuthenticatorBridgeManagerImpl.kt

```diff
@@ -0,0 +1,226 @@
+
+        // Ensure we are using the correct symmetric key:
+        val localKeyFingerprint =
+            symmetricKeyStorageProvider.symmetricKey?.toFingerprint()?.getOrNull()
+
+        // Query bridge service to see if we have a matching symmetric key:
+        val haveCorrectKey = bridgeService
+            .safeCall { checkSymmetricEncryptionKeyFingerprint(localKeyFingerprint) }
+            .fold(
+                onSuccess = { it },
+                onFailure = { false },
+            )
+            ?: false
+
+        if (!haveCorrectKey) {
+            // If we don't have the correct key, query for key:
+            bridgeService
+                .safeCall { symmetricEncryptionKeyData }
+                .fold(
+                    onSuccess = {
+                        symmetricKeyStorageProvider.symmetricKey = it
+                    },
+                    onFailure = {
+                        mutableSharedAccountsStateFlow.value = AccountSyncState.Error
+                        unbindService()
+                        return
+                    },
+                )
+
+        }
+
+        if (symmetricKeyStorageProvider.symmetricKey == null) {
+            // This means bridgeService returned a null key, which means we can make
+            // no valid operations. We should disconnect form the service and expose to the
+            // calling application that authenticator sync is not enabled.
+            mutableSharedAccountsStateFlow.value = AccountSyncState.SyncNotEnabled
+            unbindService()
+            return
+        }
+
+        // Register callback:
+        bridgeService.safeCall { registerBridgeServiceCallback(authenticatorBridgeCallback) }
+
+        // Sync data:
+        bridgeService.safeCall { syncAccounts() }
+    }
+
+    private fun onServiceDisconnected() {
+        bridgeService = null
+    }
+
+    private fun unbindService() {
+        bridgeService.safeCall { unregisterBridgeServiceCallback(authenticatorBridgeCallback) }
+        bridgeService = null
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            applicationContext.unbindService(bridgeServiceConnection)
+        } catch (_: Exception) {
+            // We want to be super safe when unbinding to assure no crashes.
+        }
+    }
+
+    private fun isBitwardenAppInstalled(): Boolean =
+        // Check to see if correct Bitwarden app is installed:
+        try {
+            applicationContext.packageManager.getPackageInfo(connectionType.toPackageName(), 0)
+            true
+        } catch (e: NameNotFoundException) {
+            false
+        }
+}
+
+/**
+ * Helper function for wrapping all calls to [IAuthenticatorBridgeService] around try catch.
+ *
+ * This is important because all calls to [IAuthenticatorBridgeService] can throw
+ * DeadObjectExceptions as well as RemoteExceptions.
+ */
+private fun <T> IAuthenticatorBridgeService?.safeCall(action: IAuthenticatorBridgeService.() -> T): Result<T?> =
```

Comment:
Just noticing that this returns `Result<T?>` instead of `Result<T>`. Can we make this so the call is on `IAuthenticatorBridgeService` instead of `IAuthenticatorBridgeService?`? We should do a local check to get a non-null value for it, rather than baking that into the helper I would think.

Reply:
Gotcha, I think this solves that: d0fbe8a743afaca40de1d88d426ee04b28786ab2